### DISCUSSION TITLE
Windows 10: Battery status script change to make it report time till fully charged

### DISCUSSION THREAD

**Discussion by nvaccessAuto** [nvaccess/nvda#14919] [Ideas]
**Reported by leonarddr on 2015-08-01 19:26**
It seems that Windows 10 has the ability to indicate how long it will take until the battery of a laptop or tablet is charged. It might be interesting to change the battery script to also report this, for example whenever the hotkey is pressed twice.

**Comment by LeonarddeR** [comment]
cc @JosephSL @derekriemer 

The time I tested this initially, it wasn't technically possible to retrieve this information. I wonder whether it is possible now. Also, I recall a ticket that is related to this and where I mentioned my tests earlier, or it must have been IRC.

**Comment by derekriemer** [comment]
It's not in the battery api. If there's another way,p lease let me know.

**Comment by PratikP1** [comment]
@derekriemer Can you possibly use ChargeRateInMilliwatts in combination with FullChargeCapacityInMilliwattHours from the Battery​ReportClass to estimate the time to charge?

Reference: https://docs.microsoft.com/en-us/uwp/api/windows.devices.power.batteryreport

Feel free to tell me if I’m off.

**Comment by Adriani90** [comment]
Is this information at all displayed for sighted people? If not, i would say this should be handled by an add-on rather than NVDA core.
cc: @seanbudd

**Comment by seanbudd** [comment]
@Adriani90 - no it is not. I agree that it is better suited to an add-on. Sighted folks are generally expected to go to "Power & Settings" to find this information, or use hardware monitoring software.

I am closing this as won't fix

**Comment by XLTechie** [comment]
@seanbudd according to this article, and to my experience with Win 10 laptops 
sometimes announcing this, the sighted can access it with a mouse hover over the 
power icon in the notification area.

https://www.cnet.com/tech/computing/find-out-how-much-time-remains-before-your-laptop-is-fully-charged/

**Comment by PratikP1** [comment]
@seanbudd I find this philosophy of, “if not available to sighted people, we’ll not implement it in core” to be short-sighted. There are many screen reading functions not readily available to sighted people. Not having resources to implement a feature and prioritizing one thing over another is a different matter. There is so much not part of the core functionality that should be like proper Excel support that prevents someone like me from completely abandoning JAWS. It’s a bit disheartening at times.

**Comment by Adriani90** [comment]
Could you give examples of information reported by NVDA which is not visible for sighted people? If there is any, this should definitely treated as bug as Nvda is per definition a screenreader which implies reading what is on the screen and not behind the screen. I agree with you as far asexcel is concerned but this is out of scope for this issue.I think it needs strong arguments why non visible information should be part of the core  and not rather handled in an  addon.Von meinem iPhone gesendetAm 10.05.2023 um 15:41 schrieb Pratik Patel ***@***.***>:﻿
@seanbudd I find this philosophy of, “if not available to sighted people, we’ll not implement it in core” to be short-sighted. There are many screen reading functions not readily available to sighted people. Not having resources to implement a feature and prioritizing one thing over another is a different matter. There is so much not part of the core functionality that should be like proper Excel support that prevents someone like me from completely abandoning JAWS. It’s a bit disheartening at times.

**Comment by PratikP1** [comment]
I can think of two examples without thinking hard: 1) Retrieving font and character information; and 2) Elements list in browsers. By its nature, a screen reader can only convey information in a single dimension with some exception. It means that what the screen reader is able to convey is never equal to what is available to sighted people. So the definition of a screen reader merely there to convey information solely available to sighted people is limited and is limiting. To me, it’s a failure of imagination. I’d say considering a name and coming up with a definition of that based only on the name limits how the piece of software serves the people whom it supports. Instead of considering this piece of software as a sole conveyance for what is on the screen, I’d argue that we need to expand the definition to say that it is an information conveying mechanism.

**Comment by Qchristensen** [comment]
Can I suggest this conversation is better had in either the user group or as a discussion?

FWIW, getting back to the original request - I would say:

1) Having the amount of battery life left (in minutes) IS arguably useful.

2) This information would be useful for everyone.  There is nothing screen reader specific to the usefulness of it.

3) That request should go to Microsoft - if they implement it as perhaps a mouseover or hint on the battery information, that would potentially benefit all Windows users.  It would then absolutely be valid to ensure that NVDA reported any information as contained in such a popup.

4) Something being "useful to users of Windows (or other software) in general" (like amount of battery life left) is not justification for including it in NVDA.  To take Pratik's two examples - font formatting information is readily available visually in programs such as Microsoft Word, hence the need to have NVDA+f report the same information to NVDA users.  The fact this also works on the web and other places that information is not visually printed is partly a convenience, but partly also, a sighted user CAN see that text is italic, or underlined, or in MS Comic Sans, so the information is "there" visually, even if it isn't explicitly written on screen.  And the elements list - I gave you on Mastodon this morning - is arguably useful for navigation.  Sure a sighted user might not immediately get a count of how many links there are on a page, but they can also flick the mouse wheel to quickly scroll down and stop when they see larger text for a heading they want.  The exact method might be different, but the end result in both cases is still to endeavour to get the user to the section of the page they want as quickly as possible.  If you do have anything else which you feel strongly about, please do create an issue or comment on the appropriate issue.

**Comment by Adriani90** [comment]
According to @XLTechie, at least on Windows 10, it is already possible to get the information "tume left until fully charged" by hovering the mouse over the appropriate area.
Luke does NVDA report this information to you then when you move the mouse to the icon?

**Comment by CyrilleB79** [comment]
@Adriani90, @Qchristensen:
Why was #5259 moved to a discussion.
As indicated by @XLTechie in https://github.com/nvaccess/nvda/issues/5259#issuecomment-1541483242, sighted people can access to the time needed until full charge hovering the battery icon in the notification area. So it's a legitimate request that the screen reader make this information easily available for blind people.
Could you please reconsider your decision?

On the other hand, I too consider off-topic the very general comments related to which information a screen reader should provide. And I agree that such comments should be made in a discussion or on a users mailing list.

**Reply by seanbudd** [reply]
>sighted people can access to the time needed until full charge hovering the battery icon in the notification area. So it's a legitimate request that the screen reader make this information easily available for blind people.
Could you please reconsider your decision?

I think some discussion was needed around implementation - however I support that it is a valid issue. It just seems that my device doesn't show or wasn't showing the time to charge.

**Comment by Adriani90** [comment]
I think it makes sense to create a new feature request, but we still need to know if NVDA reports the information when hovering the mouse over the notification area. My system doesn't show this information either, at least I cannot retrieve it.

@XLTechie does NVDA report this when you hover the mouse?

If yes, then we can open  a new feature request to add this to the nvda+shift+b script and reference there this discussion. This might be easier to follow for someone who wants to implement it.

If no, then we can open a bug report for implementing the report of time to charge on mouse movement first and then also add it to the nvda+shift+b script.

**Comment by XLTechie** [comment]
@Adriani90 I have definitely heard this announced before when accessing the 
battery icon. However, I do not recall what I have done to provoke it, or where 
the mouse pointer was. Some testing is in order.

**Comment by Qchristensen** [comment]
Just testing here:

I unplugged my cable and used it on battery for awhile.  Now, when I hover over the battery icon with the mouse, the popup information reads:
Battery status: 86% remaining
2h 10min

At this point, NVDA reads the same information.

If I press NVDA+shift+b, NVDA reads:
86%, 2 hours 10 and 10 minutes remaining, AC Disconnected.

Apologies for not testing that more previously - I didn't have the cable unplugged when I checked and of course it won't tell you the time remaining then...

Now, the original issue here is against Windows 10, and I am using Windows 11 (64-bit) Version: 22H2, Build: 22621.1702 

So, this isn't an issue on Windows 11, but it would be good to confirm exactly what is presented to the user with Windows 10

**Comment by CyrilleB79** [comment]
On my computer DELL (French) with NVDA 2023.1 on Windows 10 21H2 (AMD64) build 19044.2846:
* I have unplugged it for a while
* Just after plugging it again:
  * NVDA+shift+B reports "93 pour cent Batterie en charge" (i.e. "93 percent Charging battery")
  * Focusing the battery icon in the notification area reports: "93% disponibles (branché)  bouton" (i.e. "93% available (plugged in)  button")
* After one minute or two, AC still plugged in:
  * NVDA+shift+B reports "94 pour cent Batterie en charge" (i.e. "94 percent Charging battery")
  * Focusing the battery icon in the notification area reports: "21 min jusqu’au chargement complet  bouton" (i.e.: ""21 min until full charge  button")

So @Adriani90, this is definitely the confirmation that the time until full charge is reported in some situations by the battery icon.

@Qchristensen, sorry, your last test is not relevant to this discussion:
* The current issue/discussion is about reporting the time until full charge when the computer is plugged in.
* Your test is about reporting the possible remaining time before the battery is fully discharged when the computer is not plugged in.

**Comment by XLTechie** [comment]
@Qchristensen when you take it down to 86% (or whatever), and plug it back in, 
then hover the mouse again: what does it say?

**Comment by Qchristensen** [comment]
I just unplugged it and used it down to about 70%.  Then I plugged it back in and kept using it.  When I hover over it now, the text reads:
Battery status: 94% available (plugged in).
NVDA reports:
94% charging battery.

It doesn't give me the minutes until charged.  If I look in the power settings, I can see it says 10 minutes until charged, but I can't see any option to change how it is reported in the notification area?

Same Windows 11 build as last time I commented -  Windows 11 (64-bit) Version: 22H2, Build: 22621.1702
